### PR TITLE
feat: add escapedName to parsed Result

### PR DESCRIPTION
**Pull request by nexdrew**
`escapedName` URI-encodes the slash for scoped names such that `@foo/bar` becomes `@foo%2fbar`. For unscoped names, `escapedName` should just match `name`, either null or otherwise.

The idea is that anywhere you would use `name` when talking to a registry, you should now use `escapedName` instead. This should eliminate the need for duplicating [lines like this](https://github.com/npm/npm/blob/v3.9.6/lib/utils/map-to-registry.js#L16) across multiple repos/packages/modules that need to parse a package name when working with a registry, which is something we do quite often among open and closed source code at npm. This little package seems an ideal place to centralize this logic.

When implementing this, I considered a few different options:

1. Using `Object.defineProperty` to add `escapedName` as a getter method that was always derived from `name`, conditionally escaping based on `scope`

    This kept things simple in terms of not needing to "follow" all the places in the logic where `name` and `scope` are set. Its drawback, however, was a hidden value when relying on `Object.prototype.toString()`, such as [here](https://github.com/npm/npm/blob/v3.9.6/lib/utils/map-to-registry.js#L53), which resulted in output like:

    ```
    Result {
      escapedName: [Getter],
      raw: '@foo/bar',
      scope: '@foo',
      name: '@foo/bar',
      rawSpec: '',
      spec: 'latest',
      type: 'tag' }
    ```

    Which is obviously less than ideal.

2. Using a wrapper or decorator function to populate `escapedName` whenever `res` is returned.

    This fixed the `toString()` problem, but since the `npa()` function can actually return in a few different places, I ended up "following" the return statements rather than where `name` and `scope` are actually set. This seemed less than ideal since the it didn't make the relationship between the derived property and the original properties obvious.

So instead I opted to align the setting of `escapedName` in close proximity to the setting of `scope` and `name` in an explicit, procedural way.

### CODE REVIEW

**Comment by nexdrew** [comment]
It seems that builds have been [failing](https://travis-ci.org/npm/npm-package-arg/builds) or [erring for Node 0.8](https://travis-ci.org/npm/npm-package-arg/jobs/123988650) for quite some time, so please don't hold that against this PR. 😃

File: npa.js [hunk 1]
```diff
@@ -2,7 +2,6 @@ var url = require("url")
 var assert = require("assert")
 var util = require("util")
 var semver = require("semver")
-var path = require("path")
```
Comment (othiym23) [inline npa.js]: To avoid confusion this should be in a separate commit (or PR; I think it may be time to `standard`ize this code base).

File: npa.js [hunk 2]
```diff
@@ -50,11 +49,16 @@ function npa (arg) {
   if (nameparse && validName(nameparse[3]) &&
       (!nameparse[2] || validName(nameparse[2]))) {
     res.name = (nameparse[1] || "") + nameparse[3]
-    if (nameparse[2])
+    if (nameparse[2]) {
       res.scope = "@" + nameparse[2]
+      res.escapedName = escapeName(res.name)
```
Comment (othiym23) [inline npa.js]: Is there a reason to not just hoist this out of the conditional and just make it

```javascript
res.escapedName = escapeName(res.name)
```
?

Doing so lets us change the escaping function without tying it too heavily to scopes, and also removes the need to deal with `else` branches forever.

**Comment by othiym23** [comment]
The CLI team is dropping support for 0.8 as of now, so I'll probably pull that from `.travis.yml` and we'll see what happens. A couple minor nits, but otherwise this looks good, and as a purely additive API, should be good to go once those nits have been addressed.

File: npa.js (hunk 1 above)
Comment (nexdrew) [inline npa.js]: Ok, good point, I will undo this.

File: npa.js (hunk 2 above)
Comment (nexdrew) [inline npa.js]: @othiym23 This is based on your comment in Slack:

> well, we need to be sure we’re only applying in scoped contexts

So I only applied escaping when a `scope` applies. But I can change this if you don't think it hurts anything. I could easily apply escaping whenever `name` is set.

I could even take it a step further and always apply escaping before returning, either similar to what I described in option 2 of OP, or maybe with a little `try .. finally` action.

**Comment by nexdrew** [comment]
@othiym23 Attempted to address your concerns. Tests pass for me locally. Let me know if you see anything else.

**Comment by othiym23** [comment]
Rebased and landed as e92fcc94d4f7a56b58e7afe179644aebf2194595. Thanks, Andrew!
